File: libagnos/python/src/transports.py
##############################################################################
# Part of the Agnos RPC Framework
#    http://agnos.sourceforge.net
#
# Copyright 2010, Tomer Filiba (tomerf@il.ibm.com; tomerfiliba@gmail.com)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##############################################################################

import socket
import signal
import time
import itertools
from select import select
from cStringIO import StringIO 
from contextlib import contextmanager
from subprocess import Popen, PIPE
from . import packers
from .utils import RLock


class TransportTimeout(IOError):
    pass

class Transport(object):
    def __init__(self, infile, outfile):
        self.infile = infile
        self.outfile = outfile
        self._rlock = RLock()
        self._wlock = RLock()
        self._read_length = -1
        self._read_pos = -1
        self._read_seq = -1
        self._write_seq = -1
        self._write_buffer = []
    
    def close(self):
        if self.infile:
            self.infile.close()
            self.infile = None
        if self.outfile:
            self.outfile.close()
            self.outfile = None
    
    def begin_read(self, timeout = None):
        if self._rlock.is_held_by_current_thread():
            raise IOError("begin_read is not reentrant")
        self._rlock.acquire()
        if timeout is not None and timeout < 0:
            timeout = 0
        rl, _, _ = select([self.infile], [], [], timeout)
        if not rl:
            self._rlock.release()
            raise TransportTimeout("no data received within %r seconds" % (timeout,))
        self._read_length = packers.Int32.unpack(self.infile)
        self._read_seq = packers.Int32.unpack(self.infile)
        self._read_pos = 0
        return self._read_seq
    
    def _assert_rlock(self):
        if not self._rlock.is_held_by_current_thread():
            raise IOError("thread must first call begin_read")
    
    def read(self, count):
        self._assert_rlock()
        if self._read_pos + count > self._read_length:
            raise EOFError("request to read more than available")
        data = self.infile.read(count)
        self._read_pos += len(data)
        return data
    
    def read_all(self):
        return self.read(self._read_length - self._read_pos)
    
    def end_read(self):
        self._assert_rlock()
        remaining = self._read_length - self._read_pos
        while remaining > 0:
            chunk = self.infile.read(min(remaining, 16*1024))
            remaining -= len(chunk)
        self._rlock.release()
    
    def begin_write(self, seq):
        if self._wlock.is_held_by_current_thread():
            raise IOError("begin_write() is not reentrant")
        self._wlock.acquire()
        self._write_seq = seq
        del self._write_buffer[:]
    
    def _assert_wlock(self):
        if not self._wlock.is_held_by_current_thread():
            raise IOError("thread must first call begin_write()")
    
    def write(self, data):
        self._assert_wlock()
        self._write_buffer.append(data)
    
    def reset(self):
        self._assert_wlock()
        del self._write_buffer[:]
    
    def end_write(self):
        self._assert_wlock()
        data = "".join(self._write_buffer)
        del self._write_buffer[:]
        if data:
            packers.Int32.pack(len(data), self.outfile)
            packers.Int32.pack(self._write_seq, self.outfile)
            self.outfile.write(data)
            self.outfile.flush()
        self._wlock.release()
    
    def cancel_write(self):
        self._assert_wlock()
        del self._write_buffer[:]
        self._wlock.release()
    
    @contextmanager
    def reading(self, timeout = None):
        yield self.begin_read(timeout)
        self.end_read()

    @contextmanager
    def writing(self, seq):
        self.begin_write(seq)
        yield
        self.end_write()


class WrappedTransport(object):
    def __init__(self, transport):
        self.transport = transport
    def close(self):
        return self.transport.close()
    def begin_read(self, timeout = None):
        return self.transport.begin_read(timeout)
    def read(self, count):
        return self.transport.read(count)
    def read_all(self):
        return self.transport.read_all()
    def begin_write(self, seq):
        return self.transport.begin_write(seq)
    def write(self, data):
        return self.transport.write(data)
    def reset(self):
        return self.transport.reset()
    def end_write(self):
        return self.transport.end_write()
    def cancel_write(self):
        return self.transport.cancel_write()
    def reading(self, timeout = None):
        return self.transport.reading(timeout)
    def writing(self, seq):
        return self.transport.writing(seq)


class SocketFile(object):
    CHUNK = 16*1024
    def __init__(self, sock, read_buffer_size = 64*1024):
        self.sock = sock
        self.sock.setblocking(False)
        self.read_buffer_size = read_buffer_size
        self.read_buffer = ""
    @classmethod
    def connect(cls, host, port):
        s = socket.socket()
        s.connect((host, port))
        return cls(s)
    def close(self):
        self.sock.close()
    def fileno(self):
        return self.sock.fileno()
    def flush(self):
        pass
    
    def _underlying_read(self, count):
        bufs = []
        upper_count = count + self.read_buffer_size
        timeout = None
        while upper_count > 0:
            rl = select([self.sock], [], [], timeout)[0]
            if not rl:
                break
            data = self.sock.recv(min(upper_count, self.CHUNK))
            if not data:
                break
            bufs.append(data)
            if timeout == 0:
                break
            count -= len(data)
            upper_count -= len(data)
            if count < 0:
                timeout = 0
        return "".join(bufs)
    
    def read(self, count):
        if len(self.read_buffer) >= count:
            data = self.read_buffer[:count]
            self.read_buffer = self.read_buffer[count:]
        else:
            req = count - len(self.read_buffer)
            data2 = self._underlying_read(req)
            data = self.read_buffer + data2[:req]
            self.read_buffer = data2[req:]
        if count > 0 and not data:
            raise EOFError()
        return data
    
    def write(self, data):
        while data:
            buf = data[:self.CHUNK]
            sent = self.sock.send(buf)
            data = data[sent:]


class SocketTransport(Transport):
    def __init__(self, sockfile):
        Transport.__init__(self, sockfile, sockfile)
    @classmethod
    def connect(cls, host, port):
        return cls(SocketFile.connect(host, port))
    @classmethod
    def from_socket(cls, sock):
        return cls(SocketFile(sock))


class ProcTransport(WrappedTransport):
    def __init__(self, proc, transport):
        WrappedTransport.__init__(self, transport)
        self.proc = proc
    
    def close(self):
        WrappedTransport.close(self)
        self.proc.send_signal(signal.SIGINT)
        tend = time.time() + 0.5
        while self.proc.poll() is None and time.time() <= tend:
            time.sleep(0.1)
        if self.proc.poll() is None:
            self.proc.terminate()
    
    @classmethod
    def from_executable(cls, filename, args = ("-m", "lib")):
        if isinstance(filename, str):
            cmdline = [filename]
        else:
            cmdline = filename 
        cmdline.extend(args)
        proc = Popen(cmdline, shell = False, stdin = PIPE, stdout = PIPE)
        return cls.from_proc(proc)

    @classmethod
    def from_proc(cls, proc):
        if proc.poll() is not None:
            raise ValueError("process terminated with exit code %r" % (proc.poll(),))
        if proc.stdout.readline().strip() != "AGNOS":
            raise ValueError("process did not start correctly")
        host = proc.stdout.readline().strip()
        port = int(proc.stdout.readline().strip())
        proc.stdout.close()
        transport = SocketTransport.connect(host, port)
        return cls(proc, transport)


class LoopbackTransport(Transport):
    def __init__(self):
        Transport.__init__(self, None, None)
        self.incoming = []
        self.outgoing = []
        self.seqgen = itertools.count(400000)
    
    def insert(self, blob, seq = None):
        """feed the transport with data to read"""
        if seq is None:
            seq = self.seqgen.next()
        self.incoming.append((seq, blob))
    def pop(self):
        """get the result of the previous end_write() operation"""
        if not self.outgoing:
            raise ValueError("pop() called before end_write()")
        return self.outgoing.pop(0)
    
    def close(self):
        self.incoming = None
        self.outgoing = None

    def begin_read(self, timeout = None):
        if self._rlock.is_held_by_current_thread():
            raise IOError("begin_read is not reentrant")
        self._rlock.acquire()
        if not self.incoming:
            self._rlock.release()
            raise IOError("begin_read() called before insert()")
        seq, data = self.incoming.pop(0)
        self.infile = StringIO(data)
        self._read_length = len(data)
        self._read_seq = seq
        self._read_pos = 0
        return self._read_seq
    
    def end_read(self):
        self._assert_rlock()
        self.infile = None
        self._rlock.release()
    
    def end_write(self):
        self._assert_wlock()
        data = "".join(self._write_buffer)
        del self._write_buffer[:]
        self.outgoing.append(data)
        self._wlock.release()


#===============================================================================
# transport factory
#===============================================================================

class TransportFactory(object):
    def accept(self):
        raise NotImplementedError()
    def close(self):
        raise NotImplementedError()


class SocketTransportFactory(TransportFactory):
    def __init__(self, port, host = "0.0.0.0", backlog = 10):
        self.sock = socket.socket()
        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.sock.bind((host, port))
        self.sock.listen(backlog)
        self.host, self.port = self.sock.getsockname()
    def accept(self):
        return SocketTransport.from_socket(self.sock.accept()[0])
    def close(self):
        self.sock.close()

File: libagnos/cpp/src/transports.cpp
//////////////////////////////////////////////////////////////////////////////
// Part of the Agnos RPC Framework
//    http://agnos.sourceforge.net
//
// Copyright 2011, International Business Machines Corp.
//                 Author: Tomer Filiba (tomerf@il.ibm.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <sstream>
#include <fstream>
#include <boost/iostreams/filtering_stream.hpp>
#include <boost/iostreams/device/back_inserter.hpp>
#include <boost/iostreams/copy.hpp>
#include <boost/iostreams/filter/zlib.hpp>
#include <iomanip>
#include <cstdlib>
#include <cstring>
#include "transports.hpp"


namespace agnos
{
	namespace transports
	{
		static std::string repr(const char * buf, size_t size)
		{
			std::ostringstream s;
			for (size_t i = 0; i < size; i++) {
				char ch = buf[i];
				if (ch == '\\') {
					s << "\\\\";
				}
				else if (ch >= 32 && ch <= 127) {
					s << ch;
				}
				else {
					s << "\\x" << std::hex << std::setw(2) << std::setfill('0') << (int)((unsigned char)ch);
				}
			}
			return s.str();
		}

		template <class T> inline string convert_to_string(const T& value)
		{
			// apparently this is the easiest and most portable way to convert
			// and integer to a string. who'd have thought.
			std::stringstream ss;
			ss << value;
			return ss.str();
		}

		std::ostream& operator<< (std::ostream& stream, const ITransport& trns)
		{
			stream << trns.to_string();
			return stream;
		}

		// adapted from http://lists.boost.org/boost-users/att-48158/IoStreams.cpp

		template<typename T> inline static void zlib_compress(const T& data, T& compressed)
		{
		    boost::iostreams::filtering_streambuf<boost::iostreams::output> out;
		    out.push(boost::iostreams::zlib_compressor());
		    out.push(boost::iostreams::back_inserter(compressed));
		    boost::iostreams::array_source src(&data[0], data.size());
		    boost::iostreams::copy(src, out);
		}

		template<typename T> inline static void zlib_decompress(const T& data, T& uncompressed)
		{
		    boost::iostreams::filtering_streambuf<boost::iostreams::input> in;
		    in.push(boost::iostreams::zlib_decompressor());
		    boost::iostreams::array_source src(&data[0], data.size());
		    in.push(src);
		    boost::iostreams::copy(in, boost::iostreams::back_inserter(uncompressed));
		}

		//////////////////////////////////////////////////////////////////////
		// BasicInputStream
		//////////////////////////////////////////////////////////////////////

		BasicInputStream::BasicInputStream() : 
#ifdef _MSC_VER
			std::istream(std::_Noinit), 
#endif
			_gcount(0)
		{
		}

		std::istream& BasicInputStream::read(char* buf, std::streamsize count)
		{
			_gcount = readn(buf, count);
			return *this;
		}

		std::streamsize BasicInputStream::gcount() const
		{
			return _gcount;
		}

		//////////////////////////////////////////////////////////////////////
		// BoundInputStream - a fixed-length input stream
		//////////////////////////////////////////////////////////////////////

		template <typename T> class BoundInputStream : public BasicInputStream
		{
		protected:
			shared_ptr<T> stream;
			std::streamsize remaining_length;
			bool skip_underlying;
			bool close_underlying;

		public:
			BoundInputStream(shared_ptr<T> stream, std::streamsize length,
					bool skip_underlying = true, bool close_underlying = false) :
				BasicInputStream(), stream(stream), remaining_length(length),
				skip_underlying(skip_underlying), close_underlying(close_underlying)
			{
				if (length < 0) {
					throw std::runtime_error("length must be >= 0");
				}
			}

			~BoundInputStream()
			{
				close();
			}

			void close()
			{
				DEBUG_LOG("BoundInputStream::close()");
				if (!stream) {
					return;
				}
				if (skip_underlying) {
					skip(-1);
				}
				if (close_underlying) {
					stream->close();
				}
				stream.reset();
			}

			std::streamsize available() const
			{
				return remaining_length;
			}

			std::streamsize readn(char* buf, std::streamsize count)
			{
				DEBUG_LOG("BoundInputStream::readn(" << count << ")");
				if (count > remaining_length) {
					THROW_FORMATTED(TransportEOFError, "request to read more bytes (" <<
						count << ") than available (" << remaining_length << ")");
				}
				stream->read(buf, count);
				_gcount = stream->gcount();
				remaining_length -= _gcount;
				return _gcount;
			}

			size_t skip(int count)
			{
				DEBUG_LOG("BoundInputStream::skip(" << count << ")");
				if (count < 0 || count > remaining_length) {
					count = remaining_length;
				}
				if (count <= 0) {
					return 0;
				}
				char buf[16*1024];
				std::streamsize total_skipped = 0;

				while (count > 0) {
					stream->read(buf, sizeof(buf));
					std::streamsize actually_read = stream->gcount();
					if (actually_read <= 0) {
						remaining_length = 0;
						break;
					}
					total_skipped += actually_read;
					count -= actually_read;
					remaining_length -= actually_read;
				}
				return total_skipped;
			}
		};

		//////////////////////////////////////////////////////////////////////
		// ZlibDecompressingStream - a zlib-decompressed input stream
		//////////////////////////////////////////////////////////////////////

		// http://www.boost.org/doc/libs/1_46_1/libs/iostreams/doc/concepts/filter.html
		template<typename T> class ZlibDecompressingStream : public BasicInputStream
		{
		protected:
			shared_ptr<T> stream;
		    boost::iostreams::filtering_stream<boost::iostreams::input> in;

		public:
			ZlibDecompressingStream(shared_ptr<T> stream) : stream(stream), in()
			{
			    in.push(boost::iostreams::zlib_decompressor());
			    in.push(*stream);
			}

			void close()
			{
				if (!stream) {
					return;
				}
				stream->close();
				stream.reset();
			}

			std::streamsize readn(char* buf, std::streamsize count)
			{
				stream->read(buf, count);
				_gcount = stream->gcount();
				return _gcount;
			}
		};

		//////////////////////////////////////////////////////////////////////
		// SocketStream (replacement for boost::ip::tcp::iostream)
		//////////////////////////////////////////////////////////////////////
		static boost::asio::io_service the_io_service;
		static tcp::resolver the_resolver(the_io_service);

		class SocketStream
		{
		protected:
			shared_ptr<tcp::socket> sock;

			SocketStream() : sock(new tcp::socket(the_io_service))
			{
			}

		public:
			SocketStream(shared_ptr<tcp::socket> sock) : sock(sock)
			{
			}

			static shared_ptr<SocketStream> connect(const string& host, const string& port)
			{
				shared_ptr<SocketStream> inst(new SocketStream());

				tcp::resolver::query query(host.c_str(), port.c_str());
				tcp::resolver::iterator endpoint_iterator = the_resolver.resolve(query);
				tcp::resolver::iterator end;
				boost::system::error_code error = boost::asio::error::host_not_found;

				while (error && endpoint_iterator != end) {
					inst->sock->close();
					inst->sock->connect(*endpoint_iterator++, error);
				}
				if (error) {
					//throw boost::system::system_error(error);
					THROW_FORMATTED(SocketTransportError, "failed connecting the socket to " <<
							host << " : " << port << ", error is " << error);
				}

				return inst;
			}

			static shared_ptr<SocketStream> connect(const string& host, unsigned short port)
			{
				return connect(host, convert_to_string(port));
			}

			static shared_ptr<SocketStream> accept(const tcp::acceptor& acceptor)
			{
				shared_ptr<SocketStream> inst(new SocketStream());
				acceptor.accept(*(inst->sock));
			}

			void close()
			{
				if (!sock) {
					return;
				}
				sock->shutdown(tcp::socket::shutdown_both);
				sock->close();
				sock.reset();
			}

			std::streamsize read(char * buf, std::streamsize count)
			{
				std::streamsize total = 0;
				boost::system::error_code error = 0;
				while (count > 0) {
					size_t bytes = sock->read_some(boost::asio::buffer(buf, count), error);
					if (error == boost::asio::error::eof) {
						 break; // Connection closed cleanly by peer.
					}
					else if (error) {
						throw boost::system::system_error(error);
					}
					count -= bytes;
					buf += bytes;
					total += bytes;
				}
				return total;
			}

			std::streamsize write(const char * buf, std::streamsize count)
			{
				std::streamsize total = 0;
				boost::system::error_code error = 0;
				while (count > 0) {
					size_t bytes = sock->write_some(boost::asio::buffer(buf, count), error);
					if (error) {
						throw boost::system::system_error(error);
					}
					count -= bytes;
					buf += bytes;
					total += bytes;
				}
				return total;
			}

		};

		//////////////////////////////////////////////////////////////////////
		// SocketTransport
		//////////////////////////////////////////////////////////////////////

		SocketTransport::SocketTransport(const string& hostname, unsigned short port) :
				wlock(), wbuf(), compbuf(), wseq(0), rlock(), instream()
		{
			// ugh! the port must be a string, or it silently (!!) fails
			string portstr = convert_to_string(port);
			sockstream = shared_ptr<tcp::iostream>(new tcp::iostream(hostname, portstr.c_str()));
			if (!sockstream->good()) {
				THROW_FORMATTED(SocketTransportError, "failed connecting the socket to " << hostname << " : " << portstr);
			}
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
		}

		SocketTransport::SocketTransport(const string& hostname, const string& port) :
				wlock(), wbuf(), wseq(0), rlock(), instream()
		{
			sockstream = shared_ptr<tcp::iostream>(new tcp::iostream(hostname, port.c_str()));
			if (!sockstream->good()) {
				THROW_FORMATTED(SocketTransportError, "failed connecting the socket to " << hostname << " : " << port);
			}
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
		}

		inline void SocketTransport::_assert_good()
		{
			if (sockstream->eof()) {
				throw TransportEOFError("socket is closed");
			}
			if (!sockstream->good()) {
				throw SocketTransportError("sockstream is fail()ed");
			}
		}

		SocketTransport::SocketTransport(shared_ptr<tcp::iostream> sockstream) :
				sockstream(sockstream),
				wlock(), wbuf(), wseq(0), rlock(), instream()
		{
			_assert_good();
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
		}

		SocketTransport::~SocketTransport()
		{
			close();
		}

		string SocketTransport::to_string() const
		{
			std::stringstream ss;
			if (sockstream) {
				ss << "<SocketTransport " << sockstream->rdbuf()->remote_endpoint().address().to_string()
					<< ":" << sockstream->rdbuf()->remote_endpoint().port() << ">";
			}
			else {
				ss << "<SocketTransport (closed)>";
			}
			return ss.str();
		}

		void SocketTransport::close()
		{
			if (sockstream.get() == NULL) {
				return;
			}
			DEBUG_LOG("SocketTransport::close");
			sockstream->rdbuf()->shutdown(tcp::socket::shutdown_both);
			sockstream->close();
			sockstream.reset();
		}

		inline int32_t SocketTransport::_read_int32()
		{
			int32_t tmp;
			sockstream->read((char*)&tmp, sizeof(tmp));
			_assert_good();
			DEBUG_LOG("R " << repr((const char*)&tmp, sizeof(tmp)));
			return ntohl(tmp);
		}

		inline void SocketTransport::_write_int32(int32_t val)
		{
			int32_t tmp = htonl(val);
			DEBUG_LOG("W " << repr((const char*)&tmp, sizeof(tmp)));
			sockstream->write((const char*)&tmp, sizeof(tmp));
			_assert_good();
		}

		int SocketTransport::begin_read()
		{
			if (rlock.is_held_by_current_thread()) {
				throw TransportError("begin_read is not reentrant");
			}
			rlock.lock();
			try {
				_assert_good();

				int rseq = _read_int32();
				DEBUG_LOG("rseq = " << rseq);

				int packet_length = _read_int32();
				DEBUG_LOG("packet_length = " << packet_length);

				int uncompressed_length = _read_int32();
				DEBUG_LOG("uncompressed_length = " << uncompressed_length);

				instream = shared_ptr<BoundInputStream<tcp::iostream> >(
						new BoundInputStream<tcp::iostream>(sockstream, packet_length, true, false));

				if (uncompressed_length > 0) {
					// wrap the instream with a ZlibDecompressingStream
					shared_ptr<ZlibDecompressingStream<BasicInputStream> > zds(
							new ZlibDecompressingStream<BasicInputStream>(instream));

					shared_ptr<BoundInputStream<ZlibDecompressingStream<BasicInputStream> > > bis(
							new BoundInputStream<ZlibDecompressingStream<BasicInputStream> >(
									zds, uncompressed_length, false, true));

					instream = bis;
				}

				return rseq;
			}
			catch (std::exception &exc) {
				rlock.unlock();
				throw;
			}
		}

		inline void SocketTransport::_assert_began_read()
		{
			if (!rlock.is_held_by_current_thread()) {
				throw TransportError("thread must first call begin_read");
			}
		}

		size_t SocketTransport::read(char * buf, size_t size)
		{
			_assert_began_read();
			_assert_good();
			size_t actually_read = instream->readn(buf, size);
			DEBUG_LOG("R (" << size << ", " << actually_read << ") " << repr(buf, actually_read));
			return actually_read;
		}

		void SocketTransport::end_read()
		{
			DEBUG_LOG("");
			_assert_began_read();
			instream->close();
			instream.reset();
			rlock.unlock();
		}

		void SocketTransport::begin_write(int32_t seq)
		{
			DEBUG_LOG("seq = " << seq);
			if (wlock.is_held_by_current_thread()) {
				throw TransportError("begin_write is not reentrant");
			}
			wlock.lock();
			_assert_good();
			wseq = seq;
			wbuf.clear();
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
		}

		inline void SocketTransport::_assert_began_write()
		{
			if (!wlock.is_held_by_current_thread()) {
				throw TransportError("thread must first call begin_write");
			}
		}

		void SocketTransport::write(const char * buf, size_t size)
		{
			DEBUG_LOG("W (" << size << ") " << repr(buf, size));
			_assert_began_write();
			_assert_good();
			wbuf.insert(wbuf.end(), buf, buf + size);
		}

		void SocketTransport::restart_write()
		{
			DEBUG_LOG("");
			_assert_began_write();
			wbuf.clear();
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
		}

		void SocketTransport::end_write()
		{
			_assert_began_write();
			_assert_good();

			if (wbuf.size() > 0) {
				if ((compression_threshold >= 0) && (wbuf.size() > (unsigned)compression_threshold)) {
					DEBUG_LOG("COMPRESSING!");
					compbuf.clear();
					compbuf.reserve(wbuf.size());
					zlib_compress(wbuf, compbuf);

					_write_int32(wseq);            // seq
					_write_int32(compbuf.size());  // packet length
					_write_int32(wbuf.size());     // uncompressed length

					DEBUG_LOG("W(" << compbuf.size() << ") " << repr(&compbuf[0], compbuf.size()));
					sockstream->write(&compbuf[0], compbuf.size());
					compbuf.clear();
				}
				else {
					_write_int32(wseq);            // seq
					_write_int32(wbuf.size());     // packet length
					_write_int32(0);               // 0 means uncompressed

					DEBUG_LOG("W(" << wbuf.size() << ") " << repr(&wbuf[0], wbuf.size()));
					sockstream->write(&wbuf[0], wbuf.size());
				}
				sockstream->flush();
				_assert_good();
				wbuf.clear();
				wbuf.reserve(DEFAULT_BUFFER_SIZE);
			}
			wlock.unlock();
		}

		void SocketTransport::cancel_write()
		{
			DEBUG_LOG("");
			_assert_began_write();
			wbuf.clear();
			wbuf.reserve(DEFAULT_BUFFER_SIZE);
			wlock.unlock();
		}

		//////////////////////////////////////////////////////////////////////
		// ProcTransport
		//////////////////////////////////////////////////////////////////////

#ifdef BOOST_PROCESS_SUPPORTED
		ProcTransport::ProcTransport(boost::process::child& proc, shared_ptr<ITransport> transport) :
				WrappedTransport(transport),
				proc(proc),
				_closed(false)
		{
		}

		ProcTransport::~ProcTransport()
		{
			close();
		}

		void ProcTransport::close()
		{
			if (_closed) {
				return;
			}
			DEBUG_LOG("ProcTransport::close");
			WrappedTransport::close();
			proc.terminate(false);
			_closed = true;
		}

		shared_ptr<ProcTransport> ProcTransport::connect(const string& executable)
		{
			vector<string> args;
			args.push_back(executable);
			args.push_back("-m");
			args.push_back("lib");
			return connect(executable, args);
		}

		shared_ptr<ProcTransport> ProcTransport::connect(const string& executable, const vector<string>& args)
		{
			boost::process::context ctx;
			ctx.stdin_behavior = boost::process::capture_stream();
			ctx.stdout_behavior = boost::process::capture_stream();
			ctx.stderr_behavior = boost::process::inherit_stream();
			ctx.environment = boost::process::self::get_environment();
			return connect(executable, args, ctx);
		}

		shared_ptr<ProcTransport> ProcTransport::connect(const string& executable, const vector<string>& args,
				const boost::process::context& ctx)
		{
			boost::process::child proc = boost::process::launch(executable, args, ctx);
			string sargs;
			for (int i = 0; i < args.size(); i++) {
				sargs += args[i] + " ";
			}
			DEBUG_LOG("launched process " << executable << ", args = " << sargs);
			return connect(proc);
		}

		shared_ptr<ProcTransport> ProcTransport::connect(boost::process::child& proc)
		{
			char buf[200];
			boost::process::pistream& stdout = proc.get_stdout();

			memset(buf, 0, sizeof(buf));
			stdout.getline(buf, sizeof(buf));
			string magic(buf);
			DEBUG_LOG("magic = " << magic);

			if (magic.compare("AGNOS") != 0) {
				throw ProcTransportError("process either failed to start or is not an agnos server");
			}

			memset(buf, 0, sizeof(buf));
			stdout.getline(buf, sizeof(buf));
			string hostname(buf);
			DEBUG_LOG("hostname = " << hostname);

			memset(buf, 0, sizeof(buf));
			stdout.getline(buf, sizeof(buf));
			string port(buf);
			DEBUG_LOG("port = " << port);

			shared_ptr<ITransport> trns(new SocketTransport(hostname, port));
			shared_ptr<ProcTransport> proctrns(new ProcTransport(proc, trns));
			return proctrns;
		}
#endif


	}
}


